### ISSUE TITLE
ENH: add keywords to control explode index

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#1751]
The behaviour of `explode` is currently different for geometry and non-geometry columns. We should add keywords allowing us to control it properly. See below for a reference:

>  No matter the column type, geopandas needs to be consistent. Since we also want backward compatibility, my suggestion is:

Should we? Of course, ideally yes. But on the short term, it might also makes sense to follow pandas behaviour when calling the pandas base method, and follow our behaviour when exploding a geometry column. That gives inconsistent behaviour depending on the column, on the short term. But I think having `df.explode("col")` (with "col" a non-geometry column) doing the same regardless of whether `df` is a DataFrame or a GeoDataFrame is also valuable?

> implement pandas ignore_index keyword for geom-based explode

This is indeed something we can implement anyway (and also seems useful anyway)

> add a new keyword controlling whether the returned index should be repeated or multiindex. First, we should default to MultiIndex but if we decide that consistency with pandas has a priority, we can raise a FutureWarning and change the default in 1.0 release.

Yes, that sounds good to me (this doesn't necessarily need to happen in this PR, though). 
Brainstorming on a keyword name: what it actually does is adding a "count" level to the MultiIndex, counting within each exploded group. A bit like pandas' groupby(..).cumcount(). So maybe something like `add_counts=True|False` (although you could also interpret that to just add the number of elements in each group, not a cumulative count ..). 
Or you could see it as indices into the original Multi object, so something like `add_indices` or `part_indices` ..

_Originally posted by @jorisvandenbossche in https://github.com/geopandas/geopandas/issues/1720#issuecomment-738330206_

xref #1140

**Comment by iboates** [comment]
Personally I believe that consistency with pandas should be the top priority because of its huge size and almost fundamental position in the Python world. Since most data science-oriented Python devs would almost certainly be very experienced with vanilla pandas I would imagine that it would be perceived as "unexpected behaviour" to them, given that geopandas is an offshoot library. There's no way I can back this up but I would imagine that a significant amount of people coming into the geopandas are these already-pandas-proficient people are googling "how to put polygons in a pandas dataframe" or something, and so I think it would be best to make that transition as smooth as possible. It would then make very little difference to people who are just GIS-inclined in the beginning and are perhaps learning geopandas and pandas in parallel, and even then would neutralize a future inconsistency. Furthermore, I believe that making an exception sets a precedent for future exceptions, and that it exceptions will over time be baked further and further into core functionality, making it difficult to actually remove them later and required more and more warning asterisks to be added to the documentation.

TL;DR: Consistency good. I vote for a FutureWarning

**Comment by Imanflow** [comment]
I came here after #1859, and I'm willing to take the task. But i have to admit I'm not so sure what you want to do. I'm one of those that @iboates mentioned that learned GeoPandas and Pandas together (And, maybe im biased cause is where i come from, but I don't have the feeling that the balance is as much as he says) 

Just to be sure, the intention is to replicate pandas' behavior, so repeating the index instead of MultiIndex, unless `ignore_index` is passed, which resets the index as my initial issue was? (I have to admit that reset_index sounds more clear to me that ignore_index, but if this is about be consistent with pandas.. )

**Comment by jorisvandenbossche** [comment]
I agree that the current default of GeoPandas is not ideal, and we should probably move towards (with warning?) pandas' behaviour. 

Now, I think there are 2 different (but related) aspects of this:

- The default in GeoPandas keeps the original index (repeated) and also *adds* a second level to the index with a counter per exploded geometry. The default in pandas is only to repeat the original index.
- Pandas added an `ignore_index=False/True` keyword to `explode()` recently (in pandas 1.1), which basically results in the same as doing a `.reset_index(drop=True)` afterwards. 

For the second aspect, this is something we can certainly already do and add to GeoPandas as well (since the default in pandas is `ignore_index=False`, so opt-in, this doesn't break any behaviour by adding that as an option to GeoPandas as well). @Imanflow interested in doing a PR for this? (this is basically #1859)

The first aspect (the default behaviour) is a bit trickier, since that would change behaviour if we move towards pandas' behaviour. But I think we probably want to do this, with a deprecation warning (although this will be quite noisy, as everybody using `explode()` will then see it ..) 
I think the main thing to discuss is: how to "name" this? 

Above I gave some ideas (`add_counts`, `add_indices`, `part_indices`, ...), but I am not yet fully convinced of any of those. Other ideas?

**Comment by Imanflow** [comment]
Yes, I will take the PR if you want to tackle that first. Also I can wait until a decision is made and go for both changes. 

Regarding the naming, what about `add_multiindex`?, I think it makes more clear and explicit what is doing. `add_count_as_multiindex` might be too long

**Comment by jorisvandenbossche** [comment]
> Yes, I will take the PR if you want to tackle that first. Also I can wait until a decision is made and go for both changes.

Since they can be done independently, let's already move forward with the easy part if you want to do a PR for that!

**Comment by Imanflow** [comment]
Actually, I was motivated and started to do both. I'm almost there, but a test is failing due to the .index.names

**Comment by Imanflow** [comment]
The already written tests work if I do add_multiindex (temporal name ofc) default to True. 

I created a new test: 

 ```
@pytest.mark.parametrize("index_name", [None, "test"])
    def test_explode_geodataframe_no_multiindex(self, index_name):
        # GH1393
        s = GeoSeries([MultiPoint([Point(1, 2), Point(2, 3)]), Point(5, 5)])
        df = GeoDataFrame({"level_1": [1, 2], "geometry": s})
        df.index.name = index_name

        test_df = df.explode(add_multiindex=False)

        expected_s = GeoSeries([Point(1, 2), Point(2, 3), Point(5, 5)])
        expected_df = GeoDataFrame({"level_1": [1, 1, 2], "geometry": expected_s})

        expected_index = Index([0, 0, 1], names=[index_name])
        expected_df = expected_df.set_index(expected_index)
        assert_frame_equal(test_df, expected_df)
```
 
 For index_names=None, it passes, but for index_names='test', it doesn't. 
 
Most confusing part of it is that it fails because de right df (expected_df) has None as index_names, so the test_df seems to be correct.
